src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -685,151 +672,175 @@ Containers::Optional<MeshData> UfbxImporter::doMesh(UnsignedInt id, UnsignedInt
 
         attributeData[attributeIndex++] = MeshAttributeData{
-            MeshAttribute::Bitangent, bitangents};
+            MeshAttribute::Tangent, tangentSets[i]};
         attributeOffset += sizeof(Vector3);
     }
 
-    for (std::size_t set = 0; set < mesh->uv_sets.count; set++) {
-        Containers::StridedArrayView1D<Vector2> bitangents{vertexData,
-            reinterpret_cast<Vector2*>(vertexData + attributeOffset),
-            vertexCount, stride};
+    for (UnsignedInt i = 0; i < tangentSetCount; ++i) {
+        bitangentSets[i] = {vertexData,
+            reinterpret_cast<Vector3*>(vertexData + attributeOffset),
+            indexCount, stride};
 
         attributeData[attributeIndex++] = MeshAttributeData{
-            MeshAttribute::TextureCoordinates, bitangents};
-        attributeOffset += sizeof(Vector2);
+            MeshAttribute::Bitangent, bitangentSets[i]};
+        attributeOffset += sizeof(Vector3);
     }
 
-    for (std::size_t set = 0; set < mesh->color_sets.count; set++) {
-        Containers::StridedArrayView1D<Color4> bitangents{vertexData,
+    for (UnsignedInt i = 0; i < colorSetCount; ++i) {
+        colorSets[i] = {vertexData,
             reinterpret_cast<Color4*>(vertexData + attributeOffset),
-            vertexCount, stride};
+            indexCount, stride};
 
         attributeData[attributeIndex++] = MeshAttributeData{
-            MeshAttribute::Color, bitangents};
+            MeshAttribute::Color, colorSets[i]};
         attributeOffset += sizeof(Color4);
     }
 
     CORRADE_INTERNAL_ASSERT(attributeIndex == attributeCount);
     CORRADE_INTERNAL_ASSERT(attributeOffset == stride);
 
-    return MeshData{MeshPrimitive::Triangles,
+    UnsignedInt dstIx = 0;
+    for(UnsignedInt faceIndex : mat.face_indices) {
+        ufbx_face face = mesh->faces[faceIndex];
+        UnsignedInt numTriangles = ufbx_triangulate_face(triangleIndices.data(), triangleIndices.size(), mesh, face);
+        UnsignedInt numIndices = numTriangles * 3;
+
+        for(UnsignedInt i = 0; i < numIndices; i++) {
+            UnsignedInt srcIx = triangleIndices[i];
+
+            positions[dstIx] = Vector3(mesh->vertex_position[srcIx]);
+            if (mesh->vertex_normal.exists)
+                normals[dstIx] = Vector3(mesh->vertex_normal[srcIx]);
+            for (UnsignedInt set = 0; set < uvSetCount; ++set)
+                uvSets[set][dstIx] = Vector2(mesh->uv_sets[set].vertex_uv[srcIx]);
+            for (UnsignedInt set = 0; set < tangentSetCount; ++set)
+                tangentSets[set][dstIx] = Vector3(mesh->uv_sets[set].vertex_tangent[srcIx]);
+            for (UnsignedInt set = 0; set < bitangentSetCount; ++set)
+                bitangentSets[set][dstIx] = Vector3(mesh->uv_sets[set].vertex_bitangent[srcIx]);
+            for (UnsignedInt set = 0; set < colorSetCount; ++set)
+                colorSets[set][dstIx] = Color4(mesh->color_sets[set].vertex_color[srcIx]);
+            dstIx++;
```

Comment:
Going back to your [comment about cursed exporters duplicating normal data](https://github.com/mosra/magnum-plugins/pull/133/files/8aabdc783d5b077dd987009d53e1333890493c80..22259a1505fe16131cda1325899dcbce45ff1b5b#r1027294582), I got an idea -- what if it'd deduplicate just the normals, produce an updated index buffer, assumed the rest is sufficiently unique, and then continued deduplicating the per-attribute index tuples alone? Or does the unnecessary duplication affect more than just normals?

I'm thinking going that way could result in less peak memory use. Just a random idea to think about, I'm fine with the initial PR doing the full expansion and then full deduplication. It's a less complicated code to test in any case.

Reply:
Did a quick look on my test data set on what exporters do:

* Blender, Maya, 3ds Max
  * Indexed: Position, UV, Color
  * Duplicated: Normal, Tangent, Binormal\* 
* ZBrush
  * Indexed: Position, UV
  * Duplicated: Color, Normal, Tangent, Binormal\*

\* yes it should be bitangent but that's what FBX calls it

I think the most memory-efficient way of doing it would be to do the deduplication in "real time" by using a hash map for the vertices. Unfortunately `std::unordered_map` is really annoying to get the data out of, these days I prefer hash maps that have a linear array of `(key, value)` tuples and the hash part "on the side". That way we could incrementally build the vertex buffer with no extra overhead. Though it does look like you're doing a similar thing in `removeDuplicatesInto()` just using a `std::unordered_map` as the lookup part. 

In theory I could add a "vertex buffer builder" sort of API to ufbx but that would be a bit unprecedented as it would be the first thing to allocate memory outside of the main function call.

Anyways now that I think about it the incremental construction does sound kind of neat, as it would be a lot more memory efficient and we could get rid of the `MeshTools` dependency. If this was my personal project I'd implement a custom "on the side" `u32->u32` map for this (pretty much port [rhmap](https://github.com/bqqbarbhg/rhmap) to a C++ container once again) but I'm assuming you'd prefer if I'd just use `std::unordered_map<>` and eat the slight inefficiency :D

Reply:
Hmm, this looks like a whole new project on its own :D

Don't want to block you on this, so what about keeping what you have here for the initial PR, and once I get to having a *proper* hashmap / better mesh tools (I share the same sentiment about STL maps, what a joke), we can consider building something more efficient. 

I feel a bit uneasy about adding complex functionality directly to the plugin (hard to test, hard to reuse, etc.), depending on `MeshTools` (or extracting it to `MeshTools`, testing it there and then just using it here) is still the preferrable solution.

Reply:
Sounds good, and I meant that I would have implemented some sort of hashmap at `Container::` level but like you said that gets a bit out of scope :D 

Keeping the PR as-is sounds good for now, but doing a simple `std::unordered_map<>` based eager vertex deduplication might be worth it at some point, maybe in a follow up PR at some point.

Reply:
The less the `std::unordered_map` gets used, the less places I'll have to clean up once I have my own :)